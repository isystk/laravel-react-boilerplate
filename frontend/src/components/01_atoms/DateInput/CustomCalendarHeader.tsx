import React, { ChangeEvent } from 'react'
import 'react-datepicker/dist/react-datepicker.css'

interface CustomCalendarHeaderProps {
  date: Date
  changeYear: (year: number) => void
  changeMonth: (month: number) => void
  decreaseMonth: () => void
  increaseMonth: () => void
  prevMonthButtonDisabled: boolean
  nextMonthButtonDisabled: boolean
}

const YEARS_FOR_DROPDOWN = 10

export const CustomCalendarHeader: React.FC<CustomCalendarHeaderProps> = ({
  date,
  changeYear,
  changeMonth,
  decreaseMonth,
  increaseMonth,
  prevMonthButtonDisabled,
  nextMonthButtonDisabled,
}) => {
  const years = [...Array(YEARS_FOR_DROPDOWN)].map(
    (_, i) => new Date().getFullYear() - i
  )

  return (
    <div className="react-datepicker__header">
      <div>
        <button
          className="react-datepicker__navigation react-datepicker__navigation--previous"
          onClick={decreaseMonth}
          disabled={prevMonthButtonDisabled}
        >
          ←
        </button>
        <div className="mx-2">
          {date.toLocaleDateString('ja-JP', { year: 'numeric', month: 'long' })}
        </div>
        <button
          className="react-datepicker__navigation react-datepicker__navigation--next"
          onClick={increaseMonth}
          disabled={nextMonthButtonDisabled}
        >
          →
        </button>
      </div>

      <div>
        <button
          className="react-datepicker__navigation react-datepicker__navigation--previous"
          onClick={decreaseMonth}
          disabled={prevMonthButtonDisabled}
        />
        <select
          value={date.getFullYear()}
          onChange={({ target: { value } }: ChangeEvent<HTMLSelectElement>) =>
            changeYear(Number(value))
          }
          className="react-datepicker__year-select mr-2 mt-2"
        >
          {years.map((year) => (
            <option key={year} value={year}>
              {year}
            </option>
          ))}
        </select>
        <select
          value={date.getMonth()}
          onChange={({ target: { value } }: ChangeEvent<HTMLSelectElement>) =>
            changeMonth(Number(value))
          }
          className="react-datepicker__month-select"
        >
          {Array.from(
            { length: 12 },
            (_, index) => new Date(date.getFullYear(), index).getMonth() + 1
          ).map((month, index) => (
            <option key={month} value={index}>
              {month}
            </option>
          ))}
        </select>
        <button
          className="react-datepicker__navigation react-datepicker__navigation--next"
          onClick={increaseMonth}
          disabled={nextMonthButtonDisabled}
        />
      </div>
    </div>
  )
}
