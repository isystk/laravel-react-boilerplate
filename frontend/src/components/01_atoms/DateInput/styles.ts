import { css, keyframes } from '@emotion/css'

export const dateInput = css``
