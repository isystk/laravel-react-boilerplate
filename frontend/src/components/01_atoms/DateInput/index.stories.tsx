import { Meta, Story } from '@storybook/react'
import React from 'react'
import DateInput, { DateInputProps } from './index'
import { useForm } from 'react-hook-form'

export default {
  title: '01_atoms/DateInput',
  component: DateInput,
} as Meta

type FormDateInputs = {
  birthday: string
}

const Template: Story = () => {
  const {
    control,
    register,
    formState: { errors },
  } = useForm<FormDateInputs>()
  const validate = {
    birthday: {
      required: '生年月日を入力してください',
    },
  }
  const props: DateInputProps = {
    name: 'birthday',
    placeholder: '生年月日',
    control,
    register,
    validate,
    errors,
  }
  return <DateInput {...props} />
}

export const Primary = Template.bind({})
Primary.storyName = 'プライマリ'
Primary.args = {}
