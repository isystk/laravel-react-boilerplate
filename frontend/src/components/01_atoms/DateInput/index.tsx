import React, { FC } from 'react'
import { ContainerProps } from 'types'
import { connect } from '@/components/hoc'
import { Controller } from 'react-hook-form'
import DatePicker from 'react-datepicker'
import 'react-datepicker/dist/react-datepicker.css'
import ja from 'date-fns/locale/ja'
import { addYears, subYears } from 'date-fns'
import * as styles from './styles'
import { CustomCalendarHeader } from '@/components/01_atoms/DateInput/CustomCalendarHeader'

/** DateInputProps Props */
export type DateInputProps = {
  small: boolean
  name
  placeholder: string
  value
  onChange
  control
  register
  validate
  errors
}
/** Presenter Props */
export type PresenterProps = DateInputProps

const DateInput: FC<PresenterProps> = ({
  small,
  placeholder,
  value,
  ...props
}) => (
  <DatePicker
    {...props}
    className={`${styles.dateInput} ${
      small ? 'p-2' : 'p-3'
    } bg-gray-200  border border-gray-200 rounded  leading-tight focus:outline-none focus:bg-white focus:border-gray-500`}
    placeholderText={placeholder}
    selected={value ? new Date(value) : null}
    data-testid="commentDateInput"
    peekNextMonth
    showMonthDropdown
    showYearDropdown
    dropdownMode="select"
    disabled={false}
    locale={ja}
    dateFormat="yyyy/MM/dd"
    autoComplete="off"
    renderCustomHeader={({
      date,
      changeYear,
      changeMonth,
      decreaseMonth,
      increaseMonth,
      prevMonthButtonDisabled,
      nextMonthButtonDisabled,
    }) => (
      <CustomCalendarHeader
        date={date}
        changeYear={changeYear}
        changeMonth={changeMonth}
        decreaseMonth={decreaseMonth}
        increaseMonth={increaseMonth}
        prevMonthButtonDisabled={prevMonthButtonDisabled}
        nextMonthButtonDisabled={nextMonthButtonDisabled}
      />
    )}
    minDate={subYears(new Date(), 10)}
    maxDate={addYears(new Date(), 10)}
  />
)

/** Presenter Component */
const DateInputPresenter: FC<PresenterProps> = ({
  name,
  onChange,
  control,
  register,
  validate,
  errors,
  ...props
}) => (
  <>
    {register ? (
      <Controller
        control={control}
        name={name}
        render={({ field: { onChange, value } }) => (
          <>
            <DateInput
              {...props}
              {...register(name, validate[name])}
              value={value}
              onChange={(date) => {
                onChange(date ? date.toISOString().substring(0, 10) : null)
              }}
            />
            {errors[name] && (
              <span className="pt-4 text-red-500">{errors[name].message}</span>
            )}
          </>
        )}
      />
    ) : (
      <DateInput
        {...props}
        onChange={(date) => {
          onChange(date ? date.toISOString().substring(0, 10) : null)
        }}
      />
    )}
  </>
)

/** Container Component */
const DateInputContainer: React.FC<
  ContainerProps<DateInputProps, PresenterProps>
> = ({ presenter, ...props }) => {
  return presenter({
    ...props,
  })
}

export default connect<DateInputProps, PresenterProps>(
  'DateInput',
  DateInputPresenter,
  DateInputContainer
)
