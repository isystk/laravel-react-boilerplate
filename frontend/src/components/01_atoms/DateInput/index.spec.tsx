import React from 'react'
import renderer from 'react-test-renderer'
import DateInput, { DateInputProps } from './index'
import { useForm } from 'react-hook-form'
import { renderHook } from '@testing-library/react-hooks'

type FormDateInputs = {
  birthday: string
}

describe('DateInput', () => {
  it('Match Snapshot', () => {
    const { result } = renderHook(() => useForm<FormDateInputs>())

    const {
      control,
      register,
      formState: { errors },
    } = result.current

    const validate = {
      birthday: {
        required: '生年月日を入力してください',
      },
    }
    const props: DateInputProps = {
      name: 'birthday',
      placeholder: '生年月日',
      control,
      register,
      validate,
      errors,
    }

    const component = renderer.create(<DateInput {...props} />)
    const tree = component.toJSON()

    expect(tree).toMatchSnapshot()
  })
})
