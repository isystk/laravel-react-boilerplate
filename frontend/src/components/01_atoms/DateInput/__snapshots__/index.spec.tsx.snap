// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DateInput Match Snapshot 1`] = `
<div
  className="react-datepicker-wrapper"
>
  <div
    className="react-datepicker__input-container"
  >
    <input
      autoComplete="off"
      className="css-0 p-3 bg-gray-200  border border-gray-200 rounded  leading-tight focus:outline-none focus:bg-white focus:border-gray-500"
      disabled={false}
      name="birthday"
      onBlur={[Function]}
      onChange={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      placeholder="生年月日"
      readOnly={false}
      type="text"
      value=""
    />
  </div>
</div>
`;
